File: src/common/web/api/ProjectFeaturesCommands.ts
import { Type } from "class-transformer";

import { Command } from "../core/messaging/Command";
import { CommandReply } from "../core/messaging/CommandReply";
import { CommandComposer } from "../core/messaging/composers/CommandComposer";
import { CommandReplyComposer } from "../core/messaging/composers/CommandReplyComposer";
import { MessageBuilder } from "../core/messaging/composers/MessageBuilder";
import { Message } from "../core/messaging/Message";
import { type ProjectFeatureID } from "../data/entities/features/ProjectFeature";
import { ProjectFeatures } from "../data/entities/features/ProjectFeatures";
import { type ProjectID } from "../data/entities/Project";

/**
 * Command to update the features (data) of a project.
 *
 * @param project_id - The ID of the project to update.
 * @param updated_features - List of all features (using their ID) to update.
 * @param features - The new features data.
 */
@Message.define("command/project/features/update")
export class UpdateProjectFeaturesCommand extends Command {
    public readonly project_id: ProjectID = 0;

    // @ts-ignore
    @Type(() => String)
    public readonly updated_features: ProjectFeatureID[] = [];
    // @ts-ignore
    @Type(() => ProjectFeatures)
    public readonly features: ProjectFeatures = new ProjectFeatures();

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        project_id: ProjectID,
        updated_features: ProjectFeatureID[],
        features: ProjectFeatures,
        chain: Message | null = null
    ): CommandComposer<UpdateProjectFeaturesCommand> {
        return messageBuilder.buildCommand(UpdateProjectFeaturesCommand, { project_id: project_id, updated_features: updated_features, features: features }, chain);
    }
}

/**
 * Reply to ``UpdateProjectFeaturesCommand``.
 *
 * @param project_id - The ID of the updated project.
 * @param updated_features - List of all updated features (using their ID).
 */
@Message.define("command/project/features/update/reply")
export class UpdateProjectFeaturesReply extends CommandReply {
    public readonly project_id: ProjectID = 0;

    // @ts-ignore
    @Type(() => String)
    public readonly updated_features: ProjectFeatureID[] = [];

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: UpdateProjectFeaturesReply,
        project_id: ProjectID,
        updated_features: ProjectFeatureID[],
        success: boolean = true,
        message: string = ""
    ): CommandReplyComposer<UpdateProjectFeaturesReply> {
        return messageBuilder.buildCommandReply(UpdateProjectFeaturesReply, cmd, success, message, { project_id: project_id, updated_features: updated_features });
    }
}

File: src/common/py/api/project/project_features_commands.py
import dataclasses
import typing

from ...core.messaging import (
    Command,
    CommandReply,
    Message,
)
from ...core.messaging.composers import (
    MessageBuilder,
    CommandComposer,
    CommandReplyComposer,
)
from ...data.entities.project import (
    ProjectID,
    Project,
)
from ...data.entities.project.features import (
    ProjectFeatureID,
)


@Message.define("command/project/features/update")
class UpdateProjectFeaturesCommand(Command):
    """
    Command to update the features (data) of a project.

    Args:
        project_id: The ID of the project to update.
        updated_features: List of all features (using their ID) to update.
        features: The new features data.

    Notes:
        Requires an ``UpdateProjectFeaturesReply`` reply.
    """

    project_id: ProjectID

    updated_features: typing.List[ProjectFeatureID] = dataclasses.field(
        default_factory=list
    )
    features: Project.Features = dataclasses.field(default_factory=Project.Features)

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        project_id: ProjectID,
        updated_features: typing.List[ProjectFeatureID],
        features: Project.Features,
        chain: Message | None = None,
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(
            UpdateProjectFeaturesCommand,
            chain,
            project_id=project_id,
            updated_features=updated_features,
            features=features,
        )


@Message.define("command/project/features/update/reply")
class UpdateProjectFeaturesReply(CommandReply):
    """
    Reply to ``UpdateProjectFeaturesCommand``.

    Args:
        project_id: The ID of the updated project.
        updated_features: List of all updated features (using their ID).
    """

    project_id: ProjectID

    updated_features: typing.List[ProjectFeatureID] = dataclasses.field(
        default_factory=list
    )

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: UpdateProjectFeaturesCommand,
        *,
        project_id: ProjectID,
        updated_features: typing.List[ProjectFeatureID],
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            UpdateProjectFeaturesReply,
            cmd,
            success,
            message,
            project_id=project_id,
            updated_features=updated_features,
        )
